### PR TITLE
Implement metric.calculate_unadaptability_metrics

### PR DESCRIPTION
**Pull request by ShinnosukeUesaka**
Branch created from msw-dev. Waiting on #31 to merge.

I am retrieving baseline metrics from wandb by searching tag. Please let me know if it is better to load from local file instead of using wandb.

1. Implemented metric.calculate_unadaptability_metrics

- load the val losses that already stored from fine-tuning the base model on wandb, then calculate the loss gap ratio

2. Logging evaluation losses during finetuning and save it to wandb

- The old script computes loss only at the end of finetune. With this change, it computes and saves the evaluation loss every few steps. These losses are required to compute loss gap ratio.

3. Implement utils.check_base_results_saved

- Check wandb if baseline result is saved. Using tag associated with the run to search.

5. Moved pretrain_score.run_benchmark to metrics.run_benchmark.

- The name pretrain does not make sense since we are running benchmark at different points in the script

### CODE REVIEW

**Comment by m-walters** [comment]
@ShinnosukeUesaka Merged PR 31

File: main.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 from omegaconf import DictConfig, OmegaConf
 import wandb
 
-from ufm import countermeasures, fine_tuning, models, pretrain_score, unadapt, utils
+from src.ufm import countermeasures, models, unadapt, utils, metrics, fine_tuning
```
Comment (mukobi) [inline main.py:7]: You shouldn't use src.ufm if you have the latest changes and did `pip install -e .`

File: main.py [hunk 2]
```diff
@@ -18,29 +18,49 @@ def main(cfg: DictConfig):
         # Convert config to dict type
         config=OmegaConf.to_container(cfg, resolve=True),
         mode="disabled" if cfg.disable_wandb else "online",
-        tags=cfg.get("tags", None),
+        tags=[cfg.basename if cfg.run_baseline else cfg.unadaptname], # when retrieving baseline result, we search for the latest run with the basename tag. (it might be better to search with run id to ensure uniqueness) 
         save_code=True,
+        name=cfg.basename if cfg.run_baseline else cfg.unadaptname,
```
Comment (mukobi) [inline main.py:23]: Why the choice to specify a name? If you want it, you should probably include a random identifier along with it so we don't just end up with a bunch of runs all called e.g. `Pruning`

File: main.py [hunk 3]
```diff
@@ -51,22 +71,26 @@ def main(cfg: DictConfig):
...
+    # Just fine-tune the base model
+    logger.info("No unadaptability methods provided")
+    logger.info("Fine-tuning and recording results for base model")
+    ft_val_losses = fine_tuning.run_fine_tune(
+        model_base, wandb.config.finetune, logger
     )
 
+    # TODO -- Does this make sense for the fine-tune-only base model situation?
+    logger.info("Calculating final unadaptability metrics")
+    metrics.calculate_unadaptability_metrics(ft_val_losses, wandb.config.basename, logger)
```
Comment (mukobi) [inline main.py:92]: Yes, because it should be very fast to calculate the loss gap ratio, and it's a useful smoke test to make sure this is 0 for the base model.

File: src/ufm/fine_tuning.py [hunk 4]
```diff
@@ -80,12 +120,17 @@ def tokenize_function(examples):
         trainer.train()
 
         logger.info("Evaluating model...")
-        eval_results = trainer.evaluate()
-        
-        return eval_results['eval_loss'] #validation loss for fine-tuning
+        #eval_results = trainer.evaluate()
+        trainer.save_model()
```
Comment (mukobi) [inline src/ufm/fine_tuning.py:124]: Does this save it to disk? We probably don't want this for now, just to keep it in memory. Would recommend either putting this within an if check for a new config flag, or else removing.

**Review by mukobi — CHANGES_REQUESTED** [review]

File: main.py [hunk 5]
```diff
@@ -28,45 +32,78 @@ def main(cfg: DictConfig):
...
+        model_base = models.HuggingFaceModel(wandb.config.model)
+        
+        logger.info("Benchmarking base model for relative pre-training performance")
+        metrics.run_benchmark(model_base, logger, "base")
+        ft_val_losses = fine_tuning.run_fine_tune(
+            model_base, wandb.config.finetune, logger
         )
+        # create a csv file to store the val losses
+        file_name = utils.get_base_finetune_eval_loss_path(wandb.config.baseline_metrics_path, wandb.config.model)
+        os.makedirs(os.path.dirname(file_name), exist_ok=True)
```
Comment (m-walters) [inline main.py:52]: please include a logger that this directory is being created/written to : )

File: main.py [hunk 6]
```diff
@@ -18,8 +21,9 @@ def main(cfg: DictConfig):
         # Convert config to dict type
         config=OmegaConf.to_container(cfg, resolve=True),
         mode="disabled" if cfg.disable_wandb else "online",
-        tags=cfg.get("tags", None),
         save_code=True,
+        tags=cfg.get("tags", None),
+        #name=cfg.basename if cfg.get('basename') else None,
```
[the remark this answers is not in the document]
Comment (ShinnosukeUesaka) [reply main.py]: @mukobi 
I think we can name the runs. It doesn't take that much effort to create a new name every run.

File: main.py [hunk 7]
```diff
@@ -28,45 +32,78 @@ def main(cfg: DictConfig):
...
-            model_unadapted, wandb.config.pretrain, countermeasures=True, logger=logger
-        )
-    else:
-        # Just fine-tune the base model
-        logger.info("No unadaptability methods provided")
-        model_unadapted = model_base
+    logger.info("Benchmarking unadapted model for relative pre-training performance")
+    metrics.run_benchmark(
+        model_unadapted, logger, "unadapted"
+    )
```
[the remark this answers is not in the document]
Comment (ShinnosukeUesaka) [reply main.py:79]: I refactored the code so that we have separate pipeline when cfg.run_baseline == True.
Line 41

File: configs/base_config.yaml [hunk 8]
```diff
@@ -1,9 +1,16 @@
...
+basename: baseline # wandb run name.
 device: ~  # "cpu" or "cuda"; defaults to cuda if available
 dry_run: false  # Dry run for debugging
 verbosity: 1  # [0, 1, 2] of increasing verbosity
 disable_wandb: false  # Disable WandB online integration
-model: MLP
-pretrained: False
+model: EleutherAI/pythia-14m
+baseline_metrics_path: baseline # Path to save metrics
+run_baseline: true # If True, calculates metric on the base model. If False, applies unadapt and calculate metric on anadapt model
```
[the remark this answers is not in the document]
Comment (ShinnosukeUesaka) [reply configs/base_config.yaml:10]: Ok I will keep it for now.

File: main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mukobi) [reply main.py]: But should include a random identifier with it so we can reference "Pruning-4042" and not just have a bunch of runs called "Pruning".

File: main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ShinnosukeUesaka) [reply main.py:86]: @m-walters

[the remark this answers is not in the document]
Comment (m-walters) [reply main.py:86]: @ShinnosukeUesaka 
I made a PR to be merged into this branch:
https://github.com/mukobi/Unadaptable-Foundation-Models/pull/44
It also resolves this comment
